fabric-client-gametest-api-v1/src/client/java/net/fabricmc/fabric/impl/client/gametest/ClientGameTestContextImpl.java

```diff
@@ -0,0 +1,314 @@
+				return current;
+			}
+
+			@Override
+			public String visitString(String key, String current) {
+				DEFAULT_GAME_OPTIONS.put(key, current);
+				return current;
+			}
+
+			@Override
+			public float visitFloat(String key, float current) {
+				DEFAULT_GAME_OPTIONS.put(key, current);
+				return current;
+			}
+
+			@Override
+			public <T> T visitObject(String key, T current, Function<String, T> decoder, Function<T, String> encoder) {
+				DEFAULT_GAME_OPTIONS.put(key, current);
+				return current;
+			}
+
+			@Override
+			public <T> void accept(String key, SimpleOption<T> option) {
+				DEFAULT_GAME_OPTIONS.put(key, option.getValue());
+			}
+		});
+	}
+
+	@Override
+	public void waitTick() {
+		ThreadingImpl.checkOnGametestThread("waitTick");
+		ThreadingImpl.runTick();
+	}
+
+	@Override
+	public void waitTicks(int ticks) {
+		ThreadingImpl.checkOnGametestThread("waitTicks");
+		Preconditions.checkArgument(ticks >= 0, "ticks cannot be negative");
+
+		for (int i = 0; i < ticks; i++) {
+			ThreadingImpl.runTick();
+		}
+	}
+
+	@Override
+	public void waitFor(Predicate<MinecraftClient> predicate) {
+		ThreadingImpl.checkOnGametestThread("waitFor");
+		Preconditions.checkNotNull(predicate, "predicate");
+		waitFor(predicate, DEFAULT_TIMEOUT);
+	}
+
+	@Override
+	public void waitFor(Predicate<MinecraftClient> predicate, int timeout) {
+		ThreadingImpl.checkOnGametestThread("waitFor");
+		Preconditions.checkNotNull(predicate, "predicate");
+
+		if (timeout == NO_TIMEOUT) {
+			while (!computeOnClient(predicate::test)) {
+				ThreadingImpl.runTick();
+			}
+		} else {
+			Preconditions.checkArgument(timeout > 0, "timeout must be positive");
+
+			for (int i = 0; i < timeout; i++) {
+				if (computeOnClient(predicate::test)) {
+					return;
+				}
+
+				ThreadingImpl.runTick();
+			}
+
+			if (!computeOnClient(predicate::test)) {
+				throw new AssertionError("Timed out waiting for predicate");
+			}
+		}
+	}
+
+	@Override
+	public void waitForScreen(@Nullable Class<? extends Screen> screenClass) {
+		ThreadingImpl.checkOnGametestThread("waitForScreen");
+
+		if (screenClass == null) {
+			waitFor(client -> client.currentScreen == null);
+		} else {
+			waitFor(client -> screenClass.isInstance(client.currentScreen));
+		}
+	}
+
+	@Override
+	public void setScreen(@Nullable Screen screen) {
+		ThreadingImpl.checkOnGametestThread("setScreen");
+		runOnClient(client -> client.setScreen(screen));
+	}
+
+	@Override
+	public void clickScreenButton(String translationKey) {
+		ThreadingImpl.checkOnGametestThread("clickScreenButton");
+		Preconditions.checkNotNull(translationKey, "translationKey");
+
+		runOnClient(client -> {
+			if (!tryClickScreenButtonImpl(client.currentScreen, translationKey)) {
+				throw new AssertionError("Could not find button '%s' in screen '%s'".formatted(
+					translationKey,
+					Nullables.map(client.currentScreen, screen -> screen.getClass().getName())
+				));
+			}
+		});
```

Comment:
Throwing the `AssertionError` on the gametest thread would produce a more useful error message by including the line number that caused the issue. It does require an extra `computeOnClient()` call to get the screen class, but seems worth it?

```suggestion
		boolean success = tryClickScreenButton(translationKey);
		if (!success) {
			throw new AssertionError("Could not find button '%s' in screen '%s'".formatted(
				translationKey,
				computeOnClient(client -> Nullables.map(client.currentScreen, screen -> screen.getClass().getName()))
			));
		}
```

Error message before:
```
java.lang.AssertionError: Could not find button 'selectWorld.invalid' in screen 'net.minecraft.client.gui.screen.world.CreateWorldScreen'
        at knot/net.fabricmc.fabric.impl.client.gametest.ClientGameTestContextImpl.lambda$clickScreenButton$4(ClientGameTestContextImpl.java:173) ~[fabric-client-gametest-api-v1-0.0.1+6c6eb17b04-dev.jar:?]
        at knot/net.fabricmc.fabric.impl.client.gametest.ClientGameTestContextImpl.lambda$runOnClient$10(ClientGameTestContextImpl.java:302) ~[fabric-client-gametest-api-v1-0.0.1+6c6eb17b04-dev.jar:?]
        at knot/net.fabricmc.fabric.impl.client.gametest.ThreadingImpl.lambda$runOnClient$1(ThreadingImpl.java:145) ~[fabric-client-gametest-api-v1-0.0.1+6c6eb17b04-dev.jar:?]
        at knot/net.minecraft.client.MinecraftClient.handler$znm001$fabric-client-gametest-api-v1$postRunTasks(MinecraftClient.java:6140) ~[minecraft-clientOnly-18fc096dbe-1.21.4-net.fabricmc.yarn.1_21_4.1.21.4+build.1-v2.jar:?]
        at knot/net.minecraft.client.MinecraftClient.render(MinecraftClient.java:1296) ~[minecraft-clientOnly-18fc096dbe-1.21.4-net.fabricmc.yarn.1_21_4.1.21.4+build.1-v2.jar:?]
        at knot/net.minecraft.client.MinecraftClient.run$mixinextras$wrapped$267(MinecraftClient.java:922) ~[minecraft-clientOnly-18fc096dbe-1.21.4-net.fabricmc.yarn.1_21_4.1.21.4+build.1-v2.jar:?]
        at knot/net.minecraft.client.MinecraftClient.mixinextras$bridge$run$mixinextras$wrapped$267$268(MinecraftClient.java) ~[minecraft-clientOnly-18fc096dbe-1.21.4-net.fabricmc.yarn.1_21_4.1.21.4+build.1-v2.jar:?]
        at knot/net.minecraft.client.MinecraftClient.wrapMethod$znm000$fabric-client-gametest-api-v1$onRun(MinecraftClient.java:6103) ~[minecraft-clientOnly-18fc096dbe-1.21.4-net.fabricmc.yarn.1_21_4.1.21.4+build.1-v2.jar:?]    
        at knot/net.minecraft.client.MinecraftClient.run(MinecraftClient.java) ~[minecraft-clientOnly-18fc096dbe-1.21.4-net.fabricmc.yarn.1_21_4.1.21.4+build.1-v2.jar:?]
        at knot/net.minecraft.client.main.Main.main(Main.java:267) ~[minecraft-clientOnly-18fc096dbe-1.21.4-net.fabricmc.yarn.1_21_4.1.21.4+build.1-v2.jar:?]
        at net.fabricmc.loader.impl.game.minecraft.MinecraftGameProvider.launch(MinecraftGameProvider.java:480) ~[fabric-loader-0.16.9.jar:?]
        at net.fabricmc.loader.impl.launch.knot.Knot.launch(Knot.java:74) ~[fabric-loader-0.16.9.jar:?]
        at net.fabricmc.loader.impl.launch.knot.KnotClient.main(KnotClient.java:23) ~[fabric-loader-0.16.9.jar:?]
        at net.fabricmc.devlaunchinjector.Main.main(Main.java:86) ~[dev-launch-injector-0.2.1+build.8.jar:?]
```

Error message after:
```
java.lang.AssertionError: Could not find button 'selectWorld.invalid' in screen 'net.minecraft.client.gui.screen.world.CreateWorldScreen'
        at knot/net.fabricmc.fabric.impl.client.gametest.ClientGameTestContextImpl.clickScreenButton(ClientGameTestContextImpl.java:173) ~[fabric-client-gametest-api-v1-0.0.1+6c6eb17b04-dev.jar:?]
        at knot/net.fabricmc.fabric.test.gametest.client.ClientGameTestTest.runTest(ClientGameTestTest.java:66) ~[testmodClient/:?]
        at knot/net.fabricmc.fabric.impl.client.gametest.FabricClientGameTestRunner.lambda$start$0(FabricClientGameTestRunner.java:48) ~[fabric-client-gametest-api-v1-0.0.1+6c6eb17b04-dev.jar:?]  
        at knot/net.fabricmc.fabric.impl.client.gametest.ThreadingImpl.lambda$runTestThread$0(ThreadingImpl.java:110) ~[fabric-client-gametest-api-v1-0.0.1+6c6eb17b04-dev.jar:?]
        at java.base/java.lang.Thread.run(Thread.java:1583) [?:?]
```

(Only the "after" message shows that `ClientGameTestTest.java:66` is where I put the invalid `clickScreenButton()` call.)

Reply:
The stacktraces not crossing the `runOnClient` boundary is a more general problem that I was going to fix later. I'm thinking of possibly changing `ThreadingImpl` to modify/add a stack trace to exceptions thrown.